spec.html

```diff
@@ -8871,18 +8871,22 @@ <h1>[[Construct]] ( _argumentsList_, _newTarget_ )</h1>
     </emu-clause>
 
     <emu-clause id="sec-createbuiltinfunction" aoid="CreateBuiltinFunction">
-      <h1>CreateBuiltinFunction ( _steps_, _internalSlotsList_ [ , _realm_ [ , _prototype_ ] ] )</h1>
-      <p>The abstract operation CreateBuiltinFunction takes arguments _steps_ and _internalSlotsList_ (a List of names of internal slots) and optional arguments _realm_ and _prototype_. _internalSlotsList_ contains the names of additional internal slots that must be defined as part of the object. This operation creates a built-in function object. It performs the following steps when called:</p>
+      <h1>CreateBuiltinFunction ( _stepsOrClosure_, _internalSlotsList_ [ , _realm_ [ , _prototype_ ] ] )</h1>
+      <p>The abstract operation CreateBuiltinFunction takes arguments _stepsOrClosure_ and _internalSlotsList_ (a List of names of internal slots) and optional arguments _realm_ and _prototype_. _internalSlotsList_ contains the names of additional internal slots that must be defined as part of the object. This operation creates a built-in function object. It performs the following steps when called:</p>
       <emu-alg>
-        1. Assert: _steps_ is either a set of algorithm steps or other definition of a function's behaviour provided in this specification.
+        1. Assert: _stepsOrClosure_ is either an Abstract Closure, a set of algorithm steps, or some other definition of a function's behaviour provided in this specification.
         1. If _realm_ is not present, set _realm_ to the current Realm Record.
         1. Assert: _realm_ is a Realm Record.
         1. If _prototype_ is not present, set _prototype_ to _realm_.[[Intrinsics]].[[%Function.prototype%]].
-        1. Let _func_ be a new built-in function object that when called performs the action described by _steps_. The new function object has internal slots whose names are the elements of _internalSlotsList_.
+        1. Let _func_ be a new built-in function object that when called performs the action described by _stepsOrClosure_. The new function object has internal slots whose names are the elements of _internalSlotsList_.
         1. Set _func_.[[Realm]] to _realm_.
         1. Set _func_.[[Prototype]] to _prototype_.
         1. Set _func_.[[Extensible]] to *true*.
         1. Set _func_.[[ScriptOrModule]] to *null*.
+        1. If _stepsOrClosure_ is an Abstract Closure, then
+          1. Let _len_ be number of parameters taken by _stepsOrClosure_.
+          1. Perform ! SetFunctionLength(_func_, _len_).
+          1. Perform ! SetFunctionName(_func_, *""*).
```

Comment:
This is odd, because it implies that, of all the built-in functions, only the ones created via a closure get `length` and `name` properties.

I'd be inclined to leave this out until/unless we have a mechanism that works for all built-in functions.

Reply:
Well, all the current built‑in functions are created by passing <var>steps</var> to `CreateBuiltinFunction` define `name` and `length` in the description of the steps.

Reply:
> Well, all the current built‑in functions are created by passing steps to `CreateBuiltinFunction`

Pretty much, yes.

> define `name` and `length` in the description of the steps.

"_steps_ is either a set of algorithm steps or other definition of a function's behaviour provided in this specification."
I don't think either possibility includes the values of the function's `name` and `length` properties.

Reply:
I am inclined to make `CreateBuiltInFunction` explicitly take `_name_` and `_length_` parameters. The function only has 20 callsites in the current spec.

Reply:
I’ll do that in its own PR and then rebase this on top of it.